int32 r_cmd
int32 l_cmd
